xcvm/lib/core/src/asset.rs

```diff
@@ -0,0 +1,215 @@
+/// Newtype for XCVM assets ID. Must be unique for each asset and must never change.
+/// This ID is an opaque, arbitrary type from the XCVM protocol and no assumption must be made on
+/// how it is computed.
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[derive(
+	Copy,
+	Clone,
+	PartialEq,
+	Eq,
+	PartialOrd,
+	Ord,
+	Debug,
+	Encode,
+	Decode,
+	TypeInfo,
+	Serialize,
+	Deserialize,
+)]
+#[repr(transparent)]
+pub struct AssetID(pub u32);
+
+impl From<AssetID> for u32 {
+	fn from(val: AssetID) -> Self {
+		val.0
+	}
+}
+
+impl From<u32> for AssetID {
+	fn from(asset: u32) -> Self {
+		AssetID(asset)
+	}
+}
+
+impl From<PICA> for AssetID {
+	fn from(_: PICA) -> Self {
+		PICA::ID
+	}
+}
+
+impl From<ETH> for AssetID {
+	fn from(_: ETH) -> Self {
+		ETH::ID
+	}
+}
+
+impl From<USDT> for AssetID {
+	fn from(_: USDT) -> Self {
+		USDT::ID
+	}
+}
+
+impl From<USDC> for AssetID {
+	fn from(_: USDC) -> Self {
+		USDC::ID
+	}
+}
+
+pub struct InvalidAsset;
+pub struct PICA;
+pub struct ETH;
+pub struct USDT;
+pub struct USDC;
+
+/// List of XCVM compatible assets.
+/// The order matter and must not be changed, adding a network on the right is safe.
+pub type Assets = (InvalidAsset, (PICA, (ETH, (USDT, (USDC, ())))));
+
+/// Type implement network must be part of [`Networks`], otherwise invalid.
+pub trait Asset {
+	const ID: AssetID;
+}
+
+impl Asset for PICA {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+impl Asset for ETH {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+impl Asset for USDT {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+impl Asset for USDC {
+	const ID: AssetID = AssetID(<Assets as IndexOf<Self, _>>::INDEX as u32);
+}
+
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[derive(
+	Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Encode, Decode, TypeInfo, Serialize, Deserialize,
+)]
+#[repr(transparent)]
+pub struct Displayed<T>(
+	#[serde(bound(serialize = "T: core::fmt::Display"))]
+	#[serde(serialize_with = "serialize_as_string")]
+	#[serde(bound(deserialize = "T: core::str::FromStr"))]
+	#[serde(deserialize_with = "deserialize_from_string")]
+	pub T,
+);
+
+fn serialize_as_string<S: Serializer, T: core::fmt::Display>(
+	t: &T,
+	serializer: S,
+) -> Result<S::Ok, S::Error> {
+	serializer.serialize_str(&t.to_string())
+}
+
+fn deserialize_from_string<'de, D: Deserializer<'de>, T: core::str::FromStr>(
+	deserializer: D,
+) -> Result<T, D::Error> {
+	let s = alloc::string::String::deserialize(deserializer)?;
+	s.parse::<T>().map_err(|_| serde::de::Error::custom("Parse from string failed"))
+}
+
+impl<T> From<T> for Displayed<T> {
+	fn from(x: T) -> Self {
+		Displayed(x)
+	}
+}
+
+#[cfg_attr(feature = "std", derive(schemars::JsonSchema))]
+#[derive(
+	Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Encode, Decode, TypeInfo, Serialize, Deserialize,
+)]
+#[serde(rename_all = "snake_case")]
+pub enum Amount {
```

Comment:
Implementing [zero trait](https://docs.rs/num/latest/num/traits/trait.Zero.html) for this might be a good idea for a quick zero check. Otherwise, you would have to use `match` and check the underlying values of the variants.

Reply:
I've moved from Fixed/Ratio to a linear equation and implemented the according operations